web/src/testgrid-data-content.ts

```diff
@@ -0,0 +1,201 @@
+  @property({ type: Boolean})
+  showTab = false;
+
+  @property({ type: String })
+  dashboardName = '';
+
+  @property({ type: String })
+  tabName?: string;
+
+
+
+  // set the functionality when any tab is clicked on
+  private onTabActivated(event: CustomEvent<{index: number}>) {
+    const tabIndex = event.detail.index;
+
+    if (tabIndex === this.activeIndex){
+      return
+    }
+
+    this.tabName = this.tabNames[tabIndex];
+    this.fetchTabGridRows();
+
+    if (this.activeIndex === 0 || tabIndex === 0){
+      this.showTab = !this.showTab;
+    }
+    this.activeIndex = tabIndex;
+    navigateTab(this.dashboardName, this.tabName)
+  }
+
+  /**
+   * Lit-element lifecycle method.
+   * Invoked when a component is added to the document's DOM.
+   */
+  connectedCallback() {
+    super.connectedCallback();
+    this.fetchTabSummariesAndNames();
+    this.fetchTabGridRows();
+  }
+
+  /**
+   * Lit-element lifecycle method.
+   * Invoked on each update to perform rendering tasks.
+   */
+  render() {
+    var tabBar = html`${
+      // make sure we only render the tabs when there are tabs
+      when(this.tabNames.length > 0, () => html`
+        <mwc-tab-bar .activeIndex=${this.activeIndex} @MDCTabBar:activated="${this.onTabActivated}">
+          ${map(
+            this.tabNames,(name: string) => html`<mwc-tab label=${name}></mwc-tab>`
+          )}
+        </mwc-tab-bar>`)
+    }`;
+    return html`
+      ${tabBar}
+      ${!this.showTab ? 
+        html`<testgrid-dashboard-summary .tabSummariesInfo=${this.tabSummariesInfo}></testgrid-dashboard-summary>` : 
+        html`${map(
+          this.tabGridRows,
+          (row: ListRowsResponse_Row) => html`<p>${row.name}</p>`
+        )}`}
+    `;
+  }
+
+  // fetch the tab summaries and tab names to populate the tab bar
+  private async fetchTabSummariesAndNames() {
+    try {
+      const response = await fetch(
+        `http://${host}/api/v1/dashboards/${this.dashboardName}/tab-summaries`
+      );
+      if (!response.ok) {
+        throw new Error(`HTTP error: ${response.status}`);
+      }
+      const data = ListTabSummariesResponse.fromJson(await response.json());
+      var tabSummaries: Array<TabSummaryInfo> = [];
+      var tabNames: string[] = ['Summary'];
+      data.tabSummaries.forEach(ts => {
+        const si = convertResponse(ts);
+        tabSummaries.push(si);
+        tabNames.push(si.name);
+      });
+      this.tabSummariesInfo = tabSummaries;
+      this.tabNames = tabNames;
+      this.highlightIndex(this.tabName);
+    } catch (error) {
+      console.error(`Could not get dashboard summaries: ${error}`);
+    }
+  }
+
+  // identify which tab to highlight on the tab bar
+  private highlightIndex(tabName: string | undefined) {
+    if (tabName === undefined){
+      return
+    }
+    var index = this.tabNames.indexOf(tabName);
+    if (index > -1){
+      this.activeIndex = index;
+    }
+  }
+
+  // fetch the tab data
+  private async fetchTabGridRows() {
```

Comment:
qq: imo the grid component should be the one handling the API call, and navigation/routing should just know 1) to get a grid component, and 2) the name of the grid to get (passed to the grid component on init)? I may be wrong on this, but in that case should the 'grid' part of navigation just show the info it'd pass for now? (e.g. the dashboard + tab name).

Reply:
That's a great suggestion and I agree 100%. I will make those changes